mindmeld/query_cache.py

```diff
@@ -14,131 +14,201 @@
-        if not os.path.isdir(self.gen_folder):
-            os.makedirs(self.gen_folder)
+        Args:
+            row_id(integer): The unique id returned by QueryCache.key_to_row_id() or
+                             QueryCache.put().
+        Returns:
+            ProcessedQuery: The ProcessedQuery associated with the identifier.
+        '''
+        cursor = self.connection.cursor()
+        cursor.execute('''
+        SELECT query FROM queries WHERE rowid=(?);
+        ''', (row_id,))
+        row = cursor.fetchone()
+        return ProcessedQuery.from_cache(json.loads(row[0]))
 
-        try:
-            # We write to a new cache temp file and then rename it to prevent file corruption
-            # due to the user cancelling the training operation midway during the
-            # file write.
-            joblib.dump(self.versioned_data, self.tmp_cache_location)
-            if os.path.isfile(self.main_cache_location):
-                os.remove(self.main_cache_location)
-            shutil.move(self.tmp_cache_location, self.main_cache_location)
-            self.is_dirty = False
-        except (OSError, IOError, KeyboardInterrupt):
-            if os.path.exists(self.main_cache_location):
-                os.remove(self.main_cache_location)
-
-            if os.path.exists(self.tmp_cache_location):
-                os.remove(self.tmp_cache_location)
-
-            logger.error(
-                "Couldn't dump query cache to disk properly, "
-                "so deleting query cache due to possible corruption."
-            )
-
-    def load(self):
-        """
-        Loads a generated query cache into memory.
-        """
-        file_location = QUERY_CACHE_PATH.format(app_path=self.app_path)
-        try:
-            versioned_data = joblib.load(file_location)
-            if "cached_queries" not in versioned_data:
-                # The old version of caching did not have versions
-                logger.warning(
-                    "The cache contains deprecated versions of queries. Please "
-                    "run this command to clear the query cache: "
-                    '"python -m <app_name> clean -q"'
-                )
-                self._cached_queries = versioned_data
-            else:
-                self._cached_queries = versioned_data["cached_queries"]
-        except (OSError, IOError, KeyboardInterrupt):
-            self._cached_queries = {}
-        self.is_dirty = False
+    def get_value(self, domain, intent, query_text):
+        row_id = self.key_to_row_id(self.get_key(domain, intent, query_text))
+        return None if row_id is None else self.get(row_id)
+
+    def get_raw_query(self, row_id):
+        '''
+        Get the raw text only from a cached example.  See notes on get().
+        '''
+        cursor = self.connection.cursor()
+        cursor.execute('''
+        SELECT raw_query FROM queries WHERE rowid=(?);
+        ''', (row_id,))
+        return cursor.fetchone()[0]
```

Comment:
Is there a risk of retuning an empty list and we referencing the index 0, creating an out of index exception?

Reply:
No, this is why I have the access method using the row_id.  It's guaranteed to be there unless there is corruption of the db file.